src/core/handshaker/security/pipelined_secure_endpoint.cc

```diff
@@ -0,0 +1,938 @@
+            {
+              grpc_core::MutexLock lock(&impl->read_queue_mu_);
+              should_unprotect = impl->ShouldContinueUnprotect();
+              if (!status.ok()) {
+                // We rely on ContinueUnprotect to fail the read if the
+                // status is not ok, since we might currently have an
+                // unprotect in progress and we want to return that data
+                // before we fail future reads.
+                impl->unprotecting_ = status;
+              } else {
+                impl->frame_protector_.TraceOp(
+                    "Read",
+                    impl->staging_protected_data_buffer_->c_slice_buffer());
+                impl->MoveStagingIntoProtectedBuffer();
+              }
+            }
+            // We now have data to unprotect so we should continue unprotecting
+            // if not currently doing so.
+            if (should_unprotect) {
+              ContinueUnprotect(std::move(impl));
+            }
+          },
+          impl->staging_protected_data_buffer_.get(), std::move(args));
+      if (read_finished_immediately) {
+        lock.Release();
+        bool should_unprotect = false;
+        {
+          grpc_core::MutexLock lock(&impl->read_queue_mu_);
+          impl->frame_protector_.TraceOp(
+              "ReadImm",
+              impl->staging_protected_data_buffer_->c_slice_buffer());
+          impl->MoveStagingIntoProtectedBuffer();
+          should_unprotect = impl->ShouldContinueUnprotect();
+        }
+        // We now have data to unprotect so we should continue unprotecting
+        // if not currently doing so.
+        if (should_unprotect) {
+          ContinueUnprotect(std::move(impl));
+        }
+      }
+    }
+
+    grpc_core::Mutex write_queue_mu_;
+    absl::StatusOr<bool> writing_ ABSL_GUARDED_BY(write_queue_mu_) = false;
+    WriteArgs last_write_args_ ABSL_GUARDED_BY(write_queue_mu_);
+    std::unique_ptr<SliceBuffer> pending_writes_
+        ABSL_GUARDED_BY(write_queue_mu_);
+    grpc_core::Mutex read_queue_mu_;
+    // We have a separate read shutdown lock in order to avoid sharing a lock
+    // between read and unprotect. This way, we can perform reads and
+    // unprotects in parallel.
+    grpc_core::Mutex shutdown_read_mu_;
+    bool waiting_for_transport_read_ ABSL_GUARDED_BY(read_queue_mu_) = false;
+    absl::StatusOr<bool> unprotecting_ ABSL_GUARDED_BY(read_queue_mu_) = false;
+    std::unique_ptr<SliceBuffer> unprotected_data_buffer_
+        ABSL_GUARDED_BY(read_queue_mu_);
+    std::unique_ptr<SliceBuffer> staging_protected_data_buffer_;
+    SliceBuffer* pending_output_buffer_ ABSL_GUARDED_BY(read_queue_mu_);
+    std::unique_ptr<SliceBuffer> protected_data_buffer_
+        ABSL_GUARDED_BY(read_queue_mu_);
+    ReadArgs last_read_args_ ABSL_GUARDED_BY(read_queue_mu_);
+    grpc_core::FrameProtector frame_protector_;
+    absl::AnyInvocable<void(absl::Status)> on_read_ = nullptr;
```

Comment:
annotate this and the remainder with lock annotations, or a comment explaining why not